### ISSUE TITLE
Merge layer + dot or cos on Tensorflow gives inconsistent output shape

### ISSUE THREAD

**Issue by moustaki** [fchollet/keras#5131]
The merge layer with dot, on TensorFlow, outputs a result that's inconsistent with its output shape:

```python
from keras.layers import Embedding, Reshape, Merge
from keras.models import Sequential
import numpy as np

left = Sequential()
left.add(Embedding(1, 2, input_length=1))
left.add(Reshape((2,)))
right = Sequential()
right.add(Embedding(2, 2, input_length=1))
right.add(Reshape((2,)))
model = Sequential()
model.add(Merge([left, right], mode='dot', dot_axes=1))

print model.output_shape
print model.predict([np.array([0,0]), np.array([0,1])])
```

Gives the following with Tensorflow:

```
Using TensorFlow backend.
(None, 1)
[[ 0.00144613 -0.00040156]
 [ 0.00144613 -0.00040156]]
```
when the expected result's shape is the following (output of the same script, but using the Theano backend):
```
Using Theano backend.
(None, 1)
[[ 0.00014675]
 [ 0.00027668]]
```

This leads to unexpected behaviors when trying to fit such layers with a batch_size > 1 on TensorFlow.

(Tested with Keras master, TF 0.12.1 and Theano 0.8.2)

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by moustaki** [comment]
I suppose this might be due to the following note on the `batch_dot` documentation:

```
With TensorFlow backend, batch_dot() only supports  ndim >= 3)
```

And indeed, adding an extra dimension seems to give the same result in both Theano and Tensorflow:

```
from keras.layers import Embedding, Reshape, Merge
from keras.models import Sequential
import numpy as np

left = Sequential()
left.add(Embedding(1, 2, input_length=1))
left.add(Reshape((2, 1)))
right = Sequential()
right.add(Embedding(2, 2, input_length=1))
right.add(Reshape((2, 1)))
model = Sequential()
model.add(Merge([left, right], mode='cos', dot_axes=1, output_shape=(1,)))
model.add(Reshape((1,)))

print model.output_shape
print model.predict([np.array([0,0]), np.array([0,1])])
```

I would suggest throwing an exception in that case. I'll submit a PR.

**Comment by bstriner** [comment]
For 2 dims, do a merge 'mul' then a lambda sum.

**Comment by ogrisel** [comment]
> I would suggest throwing an exception in that case. I'll submit a PR

Keras could just reshape the data to insert an extra 1 dimension where required and then reshape the output back to 2D, no?

**Comment by moustaki** [comment]
@ogrisel Agreed, but worth making sure it doesn't silently return unexpected results in the meantime.

@bstriner Thanks - will give it a shot to support the 2D case in batch_dot

**Comment by ogrisel** [comment]
> Agreed, but worth making sure it doesn't silently return unexpected results in the meantime.

100% agree. I and a colleague just wasted a couple of hours because of this until I tried to update keras to get the exception raised by #5132 and arrived here.

**Comment by moustaki** [comment]
OK, this is now fixed with #5280 merged (which adds support for batch_dot and ndim == 2 in TF). Closing now.

**Comment by ogrisel** [comment]
Thanks @moustaki!
